### PR TITLE
j2template-module-fixup

### PR DESCRIPTION
**Pull request by maddenp-noaa**
**Description**

While working on the modal CLI, I discovered that this module could use some code-quality cleanup. This might happen in some other modules that interact with the CLI (and with unit tests that also need to be updated) so I thought that, rather than doing one big PR, I could do smaller PRs with some of the preparatory work to keep the changes more isolated and easier to understand.

This PR updates the `uwtools.j2template` module with:

- Delegation of `J2Template` class-level docstring information to the class method docstrings
- Addition of missing type annotations in methods and functions
- Addition of param/return/raises annotations in method/function docstrings
- A separation (both via in-module code organization and function/method/variable names) of public and private members

With the exception of naming updates for now-private members, all existing unit tests pass without modification, so this is functionally a no-op.

**How Has This Been Tested?**

Passes `make test` in a dev shell and GitHub Actions.

**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation.  I have made corresponding changes to the documentation **NA**
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published **NA**

### CODE REVIEW

File: src/uwtools/j2template.py [hunk 1]
```diff
@@ -1,64 +1,20 @@
...
-    """
-
-    j2env.filters["path_join"] = path_join
-
-
-def path_join(arg):
-    """
-    A Jinja2 filter definition for joining paths.
-    """
-    return os.path.join(*arg)
```
Comment (maddenp-noaa) [inline src/uwtools/j2template.py:26]: These functions moved down later in the module...

File: src/uwtools/j2template.py [hunk 2]
```diff
@@ -1,64 +1,20 @@
...
-
-    _load_string()
-        Wrapper around loading a Jinja2 template from a string
-
-    dump_file(output_path)
-        Wrapper around writing the template to an output path
-
-    validate_config()
-        Checks to ensure that the provided configure_obj is sufficient
-        to meet the needs of the undeclared_variables in the template
```
Comment (maddenp-noaa) [inline src/uwtools/j2template.py:61]: Most of this docstring has migrated into the individual methods.

File: src/uwtools/j2template.py [hunk 3]
```diff
@@ -69,98 +25,105 @@ def __init__(
...
-        Keyword Arguments
-        -----------------
-        loader_args : dict
-            A dictionary of arguments to pass to the J2 loader
-        log_name : str
-            The name of the logging object to be used.
+        :param configure_obj: Key/value pairs needed to render the provided template.
+        :param template_path: Path to a Jinja2 template file.
+        :param template_str: An in-memory Jinja2 template.
+        :raises RuntimeError: If neither a template file or path is provided.
```
Comment (maddenp-noaa) [inline src/uwtools/j2template.py:31]: As elsewhere below, Sphinx-style param/raises/return entries have been added to methods/functions, or existing entries updated for this style. We'll be working in the future on extending this and automatically generating HTML docs from these annotations.

File: src/uwtools/j2template.py [hunk 4]
```diff
@@ -69,98 +25,105 @@ def __init__(
...
-
-        self.configure_obj = configure_obj
-        self.template_path = template_path
-        self.template_str = template_str
-        self.loader_args = kwargs.get("loader_args", {})
-
+        self._configure_obj = configure_obj
+        self._template_path = template_path
+        self._template_str = template_str
+        self._loader_args = kwargs.get("loader_args", {})
```
Comment (maddenp-noaa) [inline src/uwtools/j2template.py:37]: Many instance variables and methods have been marked private.

**Review by christinaholtNOAA — APPROVED** [review]
LGTM

**Review by NaureenBharwaniNOAA — APPROVED** [review]
After taking a look at the j2template file with you earlier, these changes make a lot of sense to me and provide consistency with other parts of our codebase. Looks good to me!

**Pull request merged**